Improve Javac(Method|Type)Binding.getKey()

datho7561 opened the pull request:
- fix binding logic for ExternalAnnotations18Test.testBug470666a
- fix binding logic for ExternalAnnotations18Test.testAnnotateMethodTypeParameter2

(posting so people can keep track of progress or want to help, not ready for merge)

Fixes #284

datho7561 committed:
[WIP] improve JavacMethodBinding.getKey()

- fix binding logic for ExternalAnnotations18Test.testBug470666a

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] fix ExternalAnnotations18Test.testAnnotateMethodTypeParameter2

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix part of testAnnotateClassTypeParameter1

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
[WIP] handle primitive types in key

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] fixed the logic for `void' types

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria commented on org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/dom/JavacTypeVariableBinding.java:
```diff
@@ -0,0 +1,30 @@
+package org.eclipse.jdt.internal.javac.dom;
```
A copyright/license header is necessary.

datho7561 committed:
[WIP] Use Type instead of TypeSymbol when making JavacTypeBinding

This should help preserve type arguments eg. List<String> preserves
String.

This change needs to also be performed for JavacMethodBinding.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] add license header to JavacTypeVariableBinding

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria commented:
Don't hesitate to merge any progress even if it's not complete: as long as it doesn't cause a regression or some structural issue, code is better being pushed in the branch.

datho7561 commented:
Okay, there is some cleanup in the code I still need to do, I'll get that done, then squash and merge.

There is still one test that's failing that Rob and I were aiming to fix. I don't yet understand why the binding key is the way it is:

for `package libs; class Lib1 <S extends Throwable, T, U extends List<T>, V> {}`, currently our portion for `U` is:

Ljava/util/List<TT;>;

whereas ECJ gives

Ljava/util/List<Llibs/Lib1;:TT;>;

datho7561 committed:
[WIP] remove the type argument parameter from JavacTypeBinding

- We are accessing it from the passed in Type instead

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] use MethodType for method bindings

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] cleanup code

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 commented:
Going to make sure this doesn't badly break javac in vscode-java, then squash and merge

datho7561 committed:
[WIP] address regression resulting in NPE

Signed-off-by: David Thompson <davthomp@redhat.com>
